cpp/src/parquet/annotations-test.cc

```diff
@@ -0,0 +1,1482 @@
+  ASSERT_TRUE(false_milliseconds_timestamp_element.__isset.logicalType);
+  ASSERT_TRUE(false_milliseconds_timestamp_element.logicalType.__isset.TIMESTAMP);
+  ASSERT_TRUE(
+      false_milliseconds_timestamp_element.logicalType.TIMESTAMP.unit.__isset.MILLIS);
+  ASSERT_FALSE(
+      false_milliseconds_timestamp_element.logicalType.TIMESTAMP.isAdjustedToUTC);
+
+  NodePtr false_microseconds_timestamp_node = PrimitiveNode::Make(
+      "false_microseconds_timestamp", Repetition::REQUIRED,
+      TimestampAnnotation::Make(false, LogicalAnnotation::TimeUnit::MicroSeconds),
+      Type::INT64);
+  format::SchemaElement false_microseconds_timestamp_element;
+  false_microseconds_timestamp_node->ToParquet(&false_microseconds_timestamp_element);
+  ASSERT_EQ(false_microseconds_timestamp_element.name, "false_microseconds_timestamp");
+  ASSERT_FALSE(false_microseconds_timestamp_element.__isset.converted_type);
+  ASSERT_TRUE(false_microseconds_timestamp_element.__isset.logicalType);
+  ASSERT_TRUE(false_microseconds_timestamp_element.logicalType.__isset.TIMESTAMP);
+  ASSERT_TRUE(
+      false_microseconds_timestamp_element.logicalType.TIMESTAMP.unit.__isset.MICROS);
+  ASSERT_FALSE(
+      false_microseconds_timestamp_element.logicalType.TIMESTAMP.isAdjustedToUTC);
+
+  NodePtr false_nanoseconds_timestamp_node = PrimitiveNode::Make(
+      "false_nanoseconds_timestamp", Repetition::REQUIRED,
+      TimestampAnnotation::Make(false, LogicalAnnotation::TimeUnit::NanoSeconds),
+      Type::INT64);
+  format::SchemaElement false_nanoseconds_timestamp_element;
+  false_nanoseconds_timestamp_node->ToParquet(&false_nanoseconds_timestamp_element);
+  ASSERT_EQ(false_nanoseconds_timestamp_element.name, "false_nanoseconds_timestamp");
+  ASSERT_FALSE(false_nanoseconds_timestamp_element.__isset.converted_type);
+  ASSERT_TRUE(false_nanoseconds_timestamp_element.__isset.logicalType);
+  ASSERT_TRUE(false_nanoseconds_timestamp_element.logicalType.__isset.TIMESTAMP);
+  ASSERT_TRUE(
+      false_nanoseconds_timestamp_element.logicalType.TIMESTAMP.unit.__isset.NANOS);
+  ASSERT_FALSE(false_nanoseconds_timestamp_element.logicalType.TIMESTAMP.isAdjustedToUTC);
+
+  NodePtr interval_node =
+      PrimitiveNode::Make("interval", Repetition::REQUIRED, IntervalAnnotation::Make(),
+                          Type::FIXED_LEN_BYTE_ARRAY, 12);
+  format::SchemaElement interval_element;
+  interval_node->ToParquet(&interval_element);
+  ASSERT_EQ(interval_element.name, "interval");
+  ASSERT_TRUE(interval_element.__isset.converted_type);
+  ASSERT_FALSE(interval_element.__isset.logicalType);
+  ASSERT_EQ(interval_element.converted_type, ToThrift(LogicalType::INTERVAL));
+
+  NodePtr int8_node = PrimitiveNode::Make("int8", Repetition::REQUIRED,
+                                          IntAnnotation::Make(8, true), Type::INT32);
+  format::SchemaElement int8_element;
+  int8_node->ToParquet(&int8_element);
+ 
```

Comment:
The tests in ShallowRoundtrip were looped up in commit f1e1b8a8eb8147d9e9f02bed3c6f2c9f5910fd44; the tests in ShallowOneway are going to need a harder think (less common code to factor out, may require preprocessor token pasting) to avoid obfuscating the intent of the tests.  I'm still working on it.